component/libfuse/native_file_io.h

```diff
@@ -0,0 +1,229 @@
+
+   Copyright © 2020-2022 Microsoft Corporation. All rights reserved.
+   Author : <blobfusedev@microsoft.com>
+
+   Permission is hereby granted, free of charge, to any person obtaining a copy
+   of this software and associated documentation files (the "Software"), to deal
+   in the Software without restriction, including without limitation the rights
+   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+   copies of the Software, and to permit persons to whom the Software is
+   furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in all
+   copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+   SOFTWARE
+*/
+
+#ifndef __NATIVE_FILE_IO_H__
+#define __NATIVE_FILE_IO_H__
+
+// Every read-write operation is counted and after N operations send a call up to update cache policy
+#define CACHE_UPDATE_COUNTER 100
+
+
+// Structure that describes file-handle object returned back to libfuse
+typedef struct {
+    uint64_t       fd;                  // Unix FD for this file
+    uint64_t       obj;                 // Handlemap.Handle object representing this handle
+    uint16_t       cnt;                 // Number of read-write operations done on this handle
+    uint8_t        dirty;               // A write operation was performed on this handle
+} file_handle_t;
+
+
+// allocate_native_file_object : Allocate a native C-struct to hold handle map object and unix FD
+static file_handle_t* allocate_native_file_object(uint64_t fd, uint64_t obj, uint64_t file_size)
+{
+    // Called on open / create calls from libfuse component
+    file_handle_t* fobj = (file_handle_t*)malloc(sizeof(file_handle_t));
+    if (fobj) {
+        memset(fobj, 0, sizeof(file_handle_t));
+        fobj->fd = fd;
+        fobj->obj = obj;
+    }
+
+    return fobj;
+}
+
+// release_native_file_object : Release the native C-struct for handle 
+static void release_native_file_object(fuse_file_info_t* fi)
+{
+    // Called on close operation from libfuse component
+    file_handle_t* handle_obj = (file_handle_t*)fi->fh;
+    if (handle_obj) {
+        free(handle_obj);
+    }
+}
+
+
+// native_pread :  Do pread on file directly without involving any Go code
+static int native_pread(char *path, char *buf, size_t size, off_t offset, file_handle_t* handle_obj)
+{
+    errno = 0;
+    int res = pread(handle_obj->fd, buf, size, offset);
+    if (res == -1)
+        res = -errno;
+    
+    #if 0
```

Comment:
Why is this behind an #if 0? Doesnt that mean the code won't be called. We should updated the cache on reads too right?

Reply:
yes "#if 0" means code is disabled. Last run that i had with this change was successful so thinking to keep it in disabled mode only.